handlers.go

```diff
@@ -125,3 +126,25 @@ type channelHandler struct{ ch *Channel }
 func (c channelHandler) Handle(ctx context.Context, call *InboundCall) {
 	c.ch.GetSubChannel(call.ServiceName()).handler.Handle(ctx, call)
 }
+
+// channelHandlerWithNative is a Handler that wraps a Channel and delegates requests
+// to SubChannels if the inbound call's service and method name are configured to
+// ignore passed-in thirdparty handlers.
+type userHandlerWithIgnore struct {
+	ch          *Channel
+	ignore      map[string]struct{} // key is serviceName::method format
+	userHandler Handler
+}
+
+func (u userHandlerWithIgnore) Handle(ctx context.Context, call *InboundCall) {
+	var sb strings.Builder
+	sb.WriteString(call.ServiceName())
+	sb.WriteString("::")
+	sb.Write(call.Method())
+
+	if _, ok := u.ignore[sb.String()]; ok {
```

Comment:
nice use of `strings.Builder`, can you double-check that it doesn't cause any allocations in doing this check?

Reply:
[source code](https://golang.org/src/strings/builder.go) says it all with unsafe package :)
```go
func (b *Builder) String() string {
	return *(*string)(unsafe.Pointer(&b.buf))
}
```

Reply:
True, but we need to confirm that the `[]byte` doesn't escape too, not just the `[]byte` --> `string` conversion.

Reply:
It seems no escape from the `Write` implementation. The `append` should not cause an escape.  Ideally we could pre-allocate a `[]byte` and use unsafe to make a string by ourselves.
```go
// Write appends the contents of p to b's buffer.
// Write always returns len(p), nil.
func (b *Builder) Write(p []byte) (int, error) {
	b.copyCheck()
	b.buf = append(b.buf, p...)
	return len(p), nil
}

// WriteByte appends the byte c to b's buffer.
// The returned error is always nil.
func (b *Builder) WriteByte(c byte) error {
	b.copyCheck()
	b.buf = append(b.buf, c)
	return nil
}

// noescape hides a pointer from escape analysis.  noescape is
// the identity function but escape analysis doesn't think the
// output depends on the input. noescape is inlined and currently
// compiles down to zero instructions.
// USE CAREFULLY!
// This was copied from the runtime; see issues 23382 and 7921.
//go:nosplit
//go:nocheckptr
func noescape(p unsafe.Pointer) unsafe.Pointer {
	x := uintptr(p)
	return unsafe.Pointer(x ^ 0)
}

func (b *Builder) copyCheck() {
	if b.addr == nil {
		// This hack works around a failing of Go's escape analysis
		// that was causing b to escape and be heap allocated.
		// See issue 23382.
		// TODO: once issue 7921 is fixed, this should be reverted to
		// just "b.addr = b".
		b.addr = (*Builder)(noescape(unsafe.Pointer(b)))
	} else if b.addr != b {
		panic("strings: illegal use of non-zero Builder copied by value")
	}
}
```

Reply:
I tried to replicate this code in a simple benchmark, and found that it did allocate,
```go
package main

import (
        "strings"
        "testing"
)

type Call struct{ Service, Method string }

var ignores = map[string]bool{"foo::bar": true}

func canHandle(c Call) bool {
        var sb strings.Builder
        sb.WriteString(c.Service)
        sb.WriteString("::")
        sb.WriteString(c.Method)

        return ignores[sb.String()]
}

func BenchmarkEscape(b *testing.B) {
        for i := 0; i < b.N; i++ {
                if canHandle(Call{"foo", "bar"}) {

                }
        }
}
```

output of benchmark:
```
BenchmarkEscape         33148530                36.5 ns/op             8 B/op          1 allocs/op
```

and pprof shows that the `strings.Builder.WriteString` is leading to the alloc:
```
(pprof) list WriteString
Total: 244MB
ROUTINE ======================== strings.(*Builder).WriteString in /home/prashant/.gimme/versions/go1.15.5.linux.amd64/src/strings/builder.go
     244MB      244MB (flat, cum)   100% of Total
         .          .    118:
         .          .    119:// WriteString appends the contents of s to b's buffer.
         .          .    120:// It returns the length of s and a nil error.
         .          .    121:func (b *Builder) WriteString(s string) (int, error) {
         .          .    122:   b.copyCheck()
     244MB      244MB    123:   b.buf = append(b.buf, s...)
         .          .    124:   return len(s), nil
         .          .    125:}
```

We're allocating a `[]byte` to write to, and I think that's being allocated on the heap rather than the stack.

Can you verify with a similar benchmark here to see if there's allocations -- if so, we may want to do this check differently.